docs/prometheus-metrics.md

```diff
@@ -0,0 +1,81 @@
+The goal is to report generic metrics (e.g. metrics related to pods/jobs/services) during the lifecycle of `JobController` so that:
+
+* Other operators built on top of it will automatically report Prometheus metrics without additional efforts;
+* It is easier for users of Kubeflow distributed training operators to monitor operator performance and behaviors using consistent set of metrics for different distributed training operators.
+
+This document outlines the list of Prometheus metrics we plan to cover in `JobController`. We follow the metric naming convention
+outlined [here](https://prometheus.io/docs/practices/naming/) and the metric types supported by Prometheus [here](https://prometheus.io/docs/concepts/metric_types/).
+
+## Pod Metrics
+
+The following metrics related to the lifecycle of pods will be reported:
+
+| Metric Name | Metric Type | Description |
+| ----------- | ------------| ----------- |
+| created_pods_total | Counter | The total number of created pods |
+| restarted_pods_total | Counter | The total number of restarted pods |
+| deleted_pods_total | Counter | The total number of deleted pods |
+| failed_pods_total | Counter | The total number of failed pods |
+
+The following metrics will be reported on each pod:
+
+| Metric Name | Metric Type | Description |
+| ----------- | ------------| ----------- |
+| container_cpu_usage_seconds_total | Counter | Cumulative cpu time consumed in seconds |
+| container_accelerator_memory_used_bytes | Gauge | Total accelerator memory allocated |
+| container_memory_usage_bytes | Gauge | Current memory usage in bytes, including all memory regardless of when it was accessed |
+| container_network_transmit_bytes_total | Counter | Cumulative count of bytes transmitted |
+| container_fs_usage_bytes | Gauge | Number of bytes that are consumed by the container on this filesystem |
+| up | Gauge | Keep-Alive check (maintained by Prometheus on its own with its `up` metric detailed in the documentation [here](https://prometheus.io/docs/concepts/jobs_instances/#automatically-generated-labels-and-time-series))) |
+
+Note that some of the above metrics are derived from [cAdvisor](https://github.com/google/cadvisor) kubelet
+integration which reports to Prometheus through our prometheus-operator installation.
+
+## Job Metrics
+
+The following metrics related to the lifecycle of jobs will be reported:
+
+| Metric Name | Metric Type | Description |
+| ----------- | ------------| ----------- |
+| created_jobs_total | Counter | The total number of created jobs |
+| deleted_jobs_total | Counter | The total number of deleted jobs |
+| completed_jobs_total | Counter | The total number of completed jobs |
+| restarted_jobs_total | Counter | The total number of restarted jobs |
+| pending_jobs_total | Gauge | The total number of pending jobs |
+| failed_jobs_total | Counter | The total number of failed jobs |
```

Comment:
@terrytangyuan  Forgot to mention this one. Do you think it is more appropriate to make this `Gauge` as well?  Do you want to represent the history failures or the current failed jobs?

Can we list the metrics label in this doc as well? This is important and useful, too. Like we can combine `pending jobs` `running jobs` and `failed jobs` into one metric job_status{status="pending/failed/running"}, WDYT?

Reply:
Let's keep it as it is for now so that the metrics are consistent for metrics with past tense v.s. metrics with present continuous tense. Currently there are no labels yet as it's hard to differentiate metrics with two different tenses and choose different metric types for those metrics.